cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -1316,6 +1316,41 @@ var waitFile = func(path string, wait time.Duration, n int) error {
 	return fmt.Errorf("no %v after waiting for %v", path, time.Duration(n)*wait)
 }
 
+// TODO: those have to be waited by udev instead
+func waitForDevice(path string) error {
+	if !osutil.FileExists(filepath.Join(dirs.GlobalRootDir, path)) {
+		pollWait := 50 * time.Millisecond
+		pollIterations := 1200
+		logger.Noticef("waiting up to %v for %v to appear", time.Duration(pollIterations)*pollWait, path)
+		if err := waitFile(filepath.Join(dirs.GlobalRootDir, path), pollWait, pollIterations); err != nil {
+			return fmt.Errorf("cannot find device: %v", err)
+		}
+	}
+	return nil
+}
+
+func getNonUEFISystemDisk(fallbacklabel string) (string, error) {
+	values, err := osutil.KernelCommandLineKeyValues("snapd_system_disk")
+	if err != nil {
+		return "", err
+	}
+	if value, ok := values["snapd_system_disk"]; ok {
+		if err := waitForDevice(value); err != nil {
+			return "", err
+		}
+		systemdDisk, err := disks.DiskFromDeviceName(value)
```

Comment:
I still wonder if we should use `DiskFromDevicePath()` to use sysfs path instead. Or maybe allow both, being `DiskFromDeviceName()` the second option. To avoid races in device names, that could be even an attack vector.

Reply:
Does the order matter? They should not conflict. Also what kind of race are we talking about. The path should also be potentially "random". Unless we accept globs.

Reply:
I think that there is actually a difference here. It is true that the sysfs path is also non-deterministic, so you could get, say
`/sys/devices/bus1/bus1path/sda` or
`/sys/devices/bus1/bus1path/sdb`
for one disk, and then for another disk
`/sys/devices/bus2/bus2path/sdb` or
`/sys/devices/bus2/bus2path/sda`
in different boots. However, the nice thing is that, if, we are searching for `/sys/devices/bus1/bus1path/sda`, if we hit the race the system will not boot instead of trying incorrectly disk 2 as its path will include `bus2/bus2path`. (to fix this maybe avoiding to specify the node would help even :slightly_smiling_face: )

But if we search for `/dev/sda`, we could easy start booting disk 2 inadvertendly, which is what I think is a security risk.

Anyway, it is true that the paths would not conflict, so no big deal. I guess I was thinking that the sysfs style should be the "recommended" one.

Reply:
But thinking more about this, shouldn't we actually allow using udev's `ID_PATH`? That would univocally identify the hw path and it is also the name of the link in `/dev/disk/by-path`, so we can very easily know which disk we are interested in.

Reply:
A nice thing is that because in the end snapd calls `udevadm` and it is fine with symlinks, using `snapd_system_disk=/dev/disk/by-path/<ID_PATH>` should work fine. Maybe that should be the recommended way of using this.

Reply:
That is a good example for documentation.